### PR TITLE
SM: 0.4.1 Fixes and Additional Objective Options

### PR DESCRIPTION
**Pull request by lordlou**
## What is this fixing or adding?
- fixed broken Item links (https://discord.com/channels/731205301247803413/1110346952442400778/1110346952442400778)
- fixed `KeyError: Flyway Right 0` exceptions cause by missing collection ordering of the loc/Items list. Also added a missing force disable of Escape Rando if no escape solution can be found. (https://discord.com/channels/731205301247803413/909599727341928538/1109248603429535834)
- fixed broken doors in some animal surprise patches (dragonimals, phantoonimals, ridleynimals and metalimals) (https://discord.com/channels/731205301247803413/1103307300787195914/1103307300787195914)
- prevent receiving items when in the first room of Ceres (message box in mode7 is broken) (https://discord.com/channels/731205301247803413/909599727341928538/1105907183268417536)
- fixed generating with "activate chozo robots" Objective (https://discord.com/channels/731205301247803413/909599727341928538/1116240299417681970)

- added save & fast reload (with Start+Select+L+R) which will save to the original starting location and reload that save, preventing any softlocks. (see https://github.com/lordlou/SMBasepatch/commit/a0f1c92bd21d9fbe17df0863ec5f4b77e180aad2) Also reverted code change that was done to prevent softlocks (mainly comeback check and forcing beam items to be local when using Door Rando)

- fixed frequent failing generation on WebHost (KeyError: 'Kraid') (https://discord.com/channels/731205301247803413/1117193442213253270/1117193442213253270)

- added "objectiveRandom", "nbObjective", "objectiveList" and adapted Objective selection options to better reflect VARIA's.

## How was this tested?
In the first 2 and 5th fixes, generations was always failing when using the reported YAML. They now succeed with the fixed applied. I also tested the YAML with a second copy of itself and 2 more different games.

For the animals surprise fix, I could repro the issues without the fix and the doors correctly transitions (isnt blocked and doesnt freeze the game) with the fix applied. It looks like this was already happening on VARIA at some point but was fixed in a recent update of the patches.

Verified that receiving items is delayed until we are not in the elevator room of Ceres.

Tested the save & reload feature and verified its working properly.

Started a WebHost locally and all generations passed the first one would fail before the fix (KeyError: 'Kraid'). Following generations succeeds after the fix.

Tested many combinations of the new "objectiveRandom", "nbObjective", "objectiveList" options.

### CODE REVIEW

**Review by ThePhar — CHANGES_REQUESTED** [review]
Some concerns I have about some of the doc comments in Objectives is they reference the internal name of the option, but on the WebHost they won't appear so it will look off.

Example:
image

This also fails to generate with default settings on WebHost as "random" is not a valid default for an `OptionSet` type Option:
image

**Comment by KitsuneZeta** [comment]
Independently of ThePhar's review, it was discovered that these changes introduced a bug where not all items get replaced with the shuffled-in items, which results in various checks not getting counted and vanilla items where an AP item is expected to be. This bug also affects Plando'd items. The specific commit that appears to be the cause of this is [a7b0c16](https://github.com/ArchipelagoMW/Archipelago/pull/1859/commits/a7b0c164eacb72747c342e6123c8b7390a504e72), as undoing that commit alone fixes the issue. These findings have already been reported to the maintainer separately.

File: worlds/sm/Options.py [hunk 1]
```diff
@@ -286,17 +290,52 @@ class Tourian(Choice):
     option_Disabled = 2
     default = 0  
 
-class Objective(OptionList):
+class ObjectiveRandom(Toggle):
+    """
+    Use randomized custom objectives. You can choose which objectives are available for the randomizer to choose from. If enabled, the randomizer 
+    will choose nbObjective objectives from objectiveList. Otherwise, only objective is used. Default is disabled.
+    """
+    display_name = "randomize custom objectives"
```
[the remark this answers is not in the document]
Comment (lordlou) [reply worlds/sm/Options.py]: Fixed with 990381068c0d6a8936781a01e9c86d333d54dcf3.

File: worlds/sm/Options.py [hunk 2]
```diff
@@ -386,6 +425,9 @@ class RelaxedRoundRobinCF(Toggle):
     "custom_preset": CustomPreset,
     "varia_custom_preset": VariaCustomPreset,
     "tourian": Tourian,
+    "objectiveRandom": ObjectiveRandom,
+    "objectiveList": ObjectiveRandomList,
+    "nbObjective": ObjectiveRandomCount,
```
[the remark this answers is not in the document]
Comment (lordlou) [reply worlds/sm/Options.py]: It was to reflect VARIA's corresponding variables but it shouldnt have been the case here. Fixed with 990381068c0d6a8936781a01e9c86d333d54dcf3.

**Comment by lordlou** [comment]
> Independently of ThePhar's review, it was discovered that these changes introduced a bug where not all items get replaced with the shuffled-in items, which results in various checks not getting counted and vanilla items where an AP item is expected to be. This bug also affects Plando'd items. The specific commit that appears to be the cause of this is [a7b0c16](https://github.com/ArchipelagoMW/Archipelago/pull/1859/commits/a7b0c164eacb72747c342e6123c8b7390a504e72), as undoing that commit alone fixes the issue. These findings have already been reported to the maintainer separately.

Fixed with da36b61ac87b67815d6bd5e5d061c041d4734149.

**Pull request merged**